docs: Octane and much polish and clarification (#935)
typed-ember/ember-cli-typescript: tests/dummy/app/templates/docs/glimmer-components.md

</aside>

Glimmer Components are defined in one of three ways: with templates only, with a template and a backing class, or with only a backing class (i.e. a `yield`-only component). When using a backing class, you get a first-class experience using TypeScript! Unfortunately, we don’t yet support type-checking for templates, but we hope to build that out eventually. Don’t let that stop you, though: types in your component classes make for a great experience, so let’s dig in and see how it works in practice.

## A simple component

A *very* simple Glimmer component which lets you change the count of a value might look like this:

```htmlbars
<button {{on "click" this.minus}}>&minus;</button>
{{this.count}}
<button {{on "click" this.plus}}>+</button>
```

```ts
import Component from '@glimmer/component';
import { tracked } from '@glimmer/tracking'
import { action } from '@ember/object';

export default class Counter extends Component {
  @tracked count = 0;
  
  @action
  plus() {
    this.count += 1;
  }
  
  @action
  minus() {
    this.count -= 1;
  }
}
```

> Notice that there are no type declarations here – but this *is* actually a well-typed component. The type of `count` is `number`, and if we accidentally wrote something like `this.count = null` the compiler would give us an error.

Reviewer: > if we accidentally wrote something like `this.count = null` the compiler would give us an error.

only if `strictNullChecks` is turned on

Author: Yeah, fair. I'll add a parenthetical somewhere explaining what the assumptions of the docs are. `strictNullChecks: true` is our *default*, so people have to explicitly opt out of it, and I'd like to keep these docs focused on the main path (while filling out answers and edge cases and so on in ancillary docs)—I don't think we should be trying to teach *TypeScript* with them, though given our audience I could see an argument that way. Along those lines, we might want a minimal TS primer that explains our configs and why we've chosen them. I'll add it to the project.

Reviewer: Do we want this documentation to only apply for apps that preserve our generated config, or is there some benefit to making it config-agnostic where possible?

Author: Tradeoffs abound. I think we can probably solve this particular one by using an example that works in both settings (assign a string instead, for example). Maybe we make good use of that kind of example and then to *also* document our norms and why, and call it out if any particular point in the docs depends on our configs.

## A component with arguments

So far so good, but of course most components aren’t quite this simple! Instead, they’re invoked by other templates and they can invoke other components themselves in their own templates.

Glimmer components can receive both *arguments* and *attributes* when they are invoked. When you are working with a component’s backing class, you have access to the arguments but *not* to the attributes. The arguments are passed to the constructor, and then available as `this.args` on the component instance afterward. Let’s imagine a component which just logs the names of its arguments when it is first constructed:

```ts
import Component from '@glimmer/component';

export default class ArgsDisplay extends Component {
  constructor(owner: unknown, args: {}) {
    super(owner, args);

    Object.keys(args).forEach(argName => {
      console.log(argName);
    });
  }
}
```

<aside>

If you’re used to the classic Ember Object model, there are two important differences in the constructor itself:

- we use `super` instead of `this._super`

> - we *must* call `super` before we do anything else

Reviewer: it might be useful to link to the part of the ECMA spec that states this as a hard requirement (I believe it is). 

This is the closest I could find
https://tc39.es/ecma262/#sec-runtime-semantics-classdefinitionevaluation

Author: Excellent suggestion, will do.

Author: Curiously, that doesn't seem to actually specify it. All the implementations throw a reference error, treating `this` as an uninitialized variable—i.e. it's a Temporal Dead Zone—but I can't find *any* reference to this in spec or other official docs.

Author: Added a note in b7e4e99; I don't love it but it's better than nothing.

Author: Resolved by 42fe7566.

</aside>

> Notice that we have to start by calling `super` with `owner` and `args`. This may be a bit different from what you’re used to in Ember or other frameworks, but is normal for sub-classes in TypeScript. If the compiler just accepted any `...arguments`, a lot of potentially *very* unsafe invocations would go through. So, instead of using `...arguments`, we explicitly pass the *specific* arguments and make sure their types match up with what the super-class expects.

Reviewer: My understanding is, that `...arguments` is only unsafe today because TS (so far) lacks support for [variadic kinds](https://github.com/Microsoft/TypeScript/issues/5453) which would give us things like typed `...rest` param tuples (https://github.com/microsoft/TypeScript/issues/5331)

If this is the case, we may want to list the known limitations, rather than describing broad and non-specific unsafe-ness (since it may one day become safe)

Author: Good point.

Author: Resolved by 42fe7566.

> The types for `owner` here and `args` line up with what the `constructor` for Glimmer components expect. The `owner` is specified as `unknown` because this is a detail we explicitly *don’t* need to know about. The `args` are `object` because a Glimmer component *always* receives an object containing its arguments, even if the caller didn’t pass anything: then it would just be an empty object.

Reviewer: > The `args` are `object`

for consistency, I would update the constructor signature to `args: object` to ensure readers aren't burdened with thinking about whether the empty interface type `{}` and the type `object` are equivalent (I myself had to check).

Author: I need to change the comment here to match the original, actually – I had updated the definition to match what the types on `Component` specify precisely to *avoid* explaining that. What this makes me realize is we probably need an `aside` linking to the explanation of the difference of the two.

> The `args` are also available on `this`, just as you’d expect, so we could change our definition to return the names of the arguments from a getter:

Reviewer: >  just as you’d expect

I'm not 100% sure that users will "just expect" this, given that if they've been working with ember components they have probably been using `this.foo` or the now deprecated `this.attrs.foo`. Do you think that readers will be better-served by treating `this.args` as a potentially new concept?

Author: I'm assuming here and throughout (per the `aside` at the top of the doc) that this is a Glimmer Components With TypeScript guide, *not* a Glimmer Components In General guide. I'd really prefer to avoid repeating the materials in the guides—but I think spots like this might be a good place to put in further links in the "Is this bit new to you? Check out *this* part of the Ember Guides" vein. I'll add something like that here and push it up.

Reviewer: Sounds like you're on top of this. I also would like to see us avoid duplicate documentation. Maybe something like 
```suggestion
The `args` passed to a Glimmer Component [are available on `this`](https://github.com/glimmerjs/glimmer.js/blob/2f840309f013898289af605abffe7aee7acc6ed5/packages/%40glimmer/component/src/component.ts#L12), so we could change our definition to return the names of the arguments from a getter:
```

Author: I've committed that for now, though I may opt to revise to a link to the Octane guides/API docs rather than the implementation in a later pass!
